kernel/src/context.rs

```diff
@@ -185,53 +186,89 @@ impl ReadCString for VmReader<'_, Fallible> {
-        let mut cloned_reader = self.clone();
-        while (buffer.len() + mem::size_of::<usize>()) <= max_len {
-            let Ok(word) = cloned_reader.read_val::<usize>() else {
-                break;
-            };
-
-            if has_zero(word) {
-                for byte in word.to_ne_bytes() {
-                    self.skip(1);
-                    buffer.push(byte);
-                    if byte == 0 {
-                        return Ok(CString::from_vec_with_nul(buffer)
-                            .expect("We provided 0 but no 0 is found"));
-                    }
+/// Reads bytes from `reader` into `buffer` until a nul byte is found.
+///
+/// This method returns the following values:
+/// 1. `Ok(true)`: If a nul byte is found in the reader;
+/// 2. `Ok(false)`: If no nul byte is found and the `reader` is exhausted;
+/// 3. `Err(_)`: If an error occurs while reading from the `reader`.
+fn read_until_nul_byte(
+    reader: &mut VmReader,
+    buffer: &mut Vec<u8>,
+    max_len: usize,
+) -> Result<bool> {
+    macro_rules! read_one_byte_at_a_time_while {
+        ($cond:expr) => {
+            while $cond {
+                let byte = reader.read_val::<u8>()?;
+                buffer.push(byte);
+                if byte == 0 {
+                    return Ok(true);
                 }
-                unreachable!("The branch should never be reached unless `has_zero` has bugs.")
             }
+        };
+    }
 
-            self.skip(size_of::<usize>());
-            buffer.extend_from_slice(&word.to_ne_bytes());
+    // Handle the first few bytes to make `cur_addr` aligned with `size_of::<usize>`
+    read_one_byte_at_a_time_while!(
+        !is_addr_aligned(reader.cursor() as usize) && buffer.len() < max_len && reader.has_remain()
+    );
+
+    // Handle the rest of the bytes in bulk
+    let mut cloned_reader = reader.clone();
+    while (buffer.len() + mem::size_of::<usize>()) <= max_len {
+        let Ok(word) = cloned_reader.read_val::<usize>() else {
+            break;
+        };
+
+        if has_zero(word) {
+            for byte in word.to_ne_bytes() {
+                reader.skip(1);
+                buffer.push(byte);
+                if byte == 0 {
+                    return Ok(true);
+                }
+            }
+            unreachable!("The branch should never be reached unless `has_zero` has bugs.")
         }
 
-        // Handle the last few bytes that are not enough for a word
-        read_one_byte_at_a_time_while!(buffer.len() < max_len);
-
-        // Maximum length exceeded before finding the null terminator
-        return_errno_with_message!(Errno::EFAULT, "Fails to read CString from user");
+        reader.skip(size_of::<usize>());
+        buffer.extend_from_slice(&word.to_ne_bytes());
     }
+
+    // Handle the last few bytes that are not enough for a word
+    read_one_byte_at_a_time_while!(buffer.len() < max_len && reader.has_remain());
+
+    Ok(false)
```

Comment:
Shouldn't this be:
```rust
    if buffer.len() >= max_len {
        return_errno_with_message!(
            Errno::EFAULT,
            "no nul terminator is present before exceeding the maximum length"
        );
    } else {
        Ok(false)
    }
```

And the trailing part of `read_cstring` becomes:
```rust
        return_errno_with_message!(
            Errno::EFAULT,
            "no nul terminator is present before reaching the buffer limit"
        );
```

They seem to be two kinds of errors, and deserve two kinds of error messages, even though the current code works fine.